libsolidity/grammar.txt

```diff
@@ -1,48 +1,96 @@
+
+Imports = StringLiteral ('as' Identifier)?
+        | ('*' | Identifier) ('as' Identifier)? 'from' StringLiteral
+        | '{' Identifier ('as' Identifier)? ( ',' Identifier ('as' Identifier)? )* '}' 'from' StringLiteral
+
+ContractPart = StateVariableDecOrDef | UsingDeclaration
+             | StructDefinition | ModifierDefinition | FunctionDefinition | EventDefinition | EnumDefinition
 
 InheritanceSpecifier = Identifier ( '(' Expression ( ',' Expression )* ')' )?
+
+StateVariableDecOrDef = TypeName ( 'public' | 'internal' | 'private' )? Identifier ('=' Expression)? ';'
+UsingDeclaration = 'using' Identifier 'for' TypeName ';'
 StructDefinition = 'struct' Identifier '{'
-                     ( VariableDeclaration (';' VariableDeclaration)* )? '}
-StateVariableDeclaration = TypeName ( 'public' | 'inheritable' | 'private' )? Identifier ';'
+                     ( VariableDeclaration ';' (VariableDeclaration ';')* )? '}'
 ModifierDefinition = 'modifier' Identifier ParameterList? Block
 FunctionDefinition = 'function' Identifier ParameterList
-					 ( Identifier | 'constant' | 'external' | 'public' | 'inheritable' | 'private' )*
-                     ( 'returns' ParameterList )? Block
+                     ( FunctionCall | Identifier | 'constant' | 'external' | 'public' | 'internal' | 'private' )*
+                     ( 'returns' (ParameterList | TypeParameterList) )? Block
+EventDefinition = 'event' Identifier (ParameterList | TypeParameterList | IndexedTypeParameterList) 'anonymous'? ';'
 
 EnumValue = Identifier
-EnumDefinition = 'enum' '{' EnumValue (',' EnumValue)* '}'
+EnumDefinition = 'enum' Identifier '{' EnumValue? (',' EnumValue)* '}'
+
+IndexedTypeParameterList = '(' ( TypeName 'indexed'? Identifier? (',' TypeName 'indexed'? Identifier?)* )? ')'
+TypeParameterList = '(' ( TypeName (',' TypeName)* )? ')'
 ParameterList = '(' ( VariableDeclaration (',' VariableDeclaration)* )? ')'
 // semantic restriction: mappings and structs (recursively) containing mappings
 // are not allowed in argument lists
-VariableDeclaration = TypeName Identifier
-TypeName = ElementaryTypeName | Identifier | Mapping | ArrayTypeName
+VariableDeclaration = TypeName Identifier | 'var' assignment
+TypeName = ElementaryTypeName | Identifier StorageLocation? | Mapping | ArrayTypeName
 Mapping = 'mapping' '(' ElementaryTypeName '=>' TypeName ')'
-ArrayTypeName = TypeName '[' (Expression)? ']'
+ArrayTypeName = TypeName StorageLocation? '[' Expression? ']'
+StorageLocation = 'memory' | 'storage'
 
 Block = '{' Statement* '}'
-Statement = IfStatement | WhileStatement | Block |
-			( Continue | Break | Return | VariableDefinition | ExpressionStatement ) ';'
+Statement = IfStatement | WhileStatement | ForStatement | Block | PlaceholderStatement |
+            ( Continue | Break | Return | Throw | VardefOrExprStmt | ExpressionStatement ) ';'
 
-ExpressionStatement = Expression
+ExpressionStatement = Expression | VariableDefinition | Assignment | 'delete' Expression
```

Comment:
`delete` is actually a regular unary operator, so it should not receive this special treatment in the grammar.